### DISCUSSION TITLE
Implementing this, directory create error and such.

### DISCUSSION THREAD

**Discussion by iBuaSpaz** [JaffaKetchup/flutter_map_tile_caching#87] [Q&A] [answered]
I feel its been quite a hasstle to make this work, but im getting far now.

I created a function in my main.dart file

`Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  FlutterMapTileCaching.initialise(await RootDirectory.normalCache);
  FMTC.instance;

  SystemChrome.setPreferredOrientations(
    [
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ],
  ).then((val) {
    runApp(
      const MyApp(),
    );
  });
}`

works fine, then in my map_control.dart file i have my map code (map and so on loads without issues)

` children: [
                TileLayer(
                  tileProvider: FMTC.instance('cache_maps').getTileProvider(),
                  urlTemplate:
                      'https://{s}.tile.thunderforest.com/{style}/{z}/{x}/{y}{r}.png?apikey={apiKey}',
                  subdomains: const ['a', 'b', 'c'],
                  additionalOptions: const {
                    'style': 'atlas',
                    'apiKey': '<apikey>',
                  },
                  userAgentPackageName: 'dev.myapp.example',
                  maxZoom: 22,
                  minZoom: 12,
                ),`

In this i have two issues:

`E/flutter (14329): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FileSystemException: Cannot create file, path = '/data/user/0/com.example.myapp/app_flutter/fmtc/stores/cache_maps/tiles/https___b.tile.thunderforest.com_atlas_14_8607_5137@2x.png_apikey=<apikey>' (OS Error: No such file or directory, errno = 2)`

So it seems it dont create a folder to store the tiles in.

second issue is, it saves the apikey as part of the image file name, im not really interested in this, is it possible to remove that? (it would not be required to have the api key i guess, to load the cache image), everything after ? should be removed and not checked on store/get the cache tile.

Any help with these two issues is appricated.

**Comment by JaffaKetchup** [comment, marked as answer]
Hi @iBuaSpaz,

Please migrate to v7, that should help solve the issues.
In terms of the query parameter (everything after '?' in the URL), in v7 it is more hidden inside the database, but can still be read. Unfortunatley, this is necessary to keep - some tile servers have the coordinates there, which make the URL unique. This uniqueness is how the algorithm locates the cached tile, it uses the exact URL and tries to locate it in the database.

Please let me know if you need help migrating!
Additionally, please let me know if you're using FMTC within a proprietary or open source application.

**Reply by iBuaSpaz** [reply]
Thanks, now i might sound new, well i am, how do i migrate to V7? i just installed this version, i thought it get the latest when using

flutter pub add flutter_map_tile_caching

**Reply by JaffaKetchup** [reply]
I updated the pub.dev listing yesterday. You just need to change the version attached to 'flutter_map_tile_caching' in your pubspec.yaml to '^7.0.0'. If you haven't distributed your application to your users yet (with the v6), you won't need to implement the migration functionality, but you may need to follow https://fmtc.jaffaketchup.dev/get-started/v6-to-v7-migration.

If you are using FMTC in a proprietary application, please let me know. It violates the GPL license I've chosen: see https://fmtc.jaffaketchup.dev/#proprietary-licensing for more information. No problem if you aren't, and if you are, I'm sure we can agree something that suits the both of us.

**Reply by iBuaSpaz** [reply]
Thanks, proprietary application means you charge money for it?

**Reply by JaffaKetchup** [reply]
Probably yes, but not necessarily. Do you share your source code with your customers? For example, do you have a GitHub repository?

**Reply by iBuaSpaz** [reply]
i do not share the intire source code no, if i have to this is kind of useless for me,
i do not charge for my apps though, and i do credit the plugins/packages used.

if this is a issue i will remove the cache system and find another solution or just use free maps like openstreetmaps.

**Reply by JaffaKetchup** [reply]
Sharing source code is **required legally** by the GPL, so you will require an alternative license. Usually these cost money from me (as explained [here](https://fmtc.jaffaketchup.dev/#proprietary-licensing)), with the only potential exception to this being if you are non-profit entirely (eg. no ads, no donations). If you make profit, please check https://fmtc.jaffaketchup.dev/is-fmtc-right-for-me to see whether it's worth it for yourself to stick with this package.

Note that although OSM tiles are 'free' in terms of cost, they are not free in terms of usage, and you must comply with their restrictions - one of those including the limitting of distributing apps with their tiles. See https://operations.osmfoundation.org/policies/tiles for more info.

**Reply by iBuaSpaz** [reply]
I know about that in OSM, but i think i will just write my own cache class then, its not that much i need, then maybe stick with thunderforest, i dont need background fetching, just cache when i user is on the map for later use, i think i saw i few simple samples for this.

It just needs to store/get/check expired when user is moving around, as thunderforest is payed.

**Reply by JaffaKetchup** [reply]
Sure, no hard feelings :). It's defintley possible to implement your own basic solution for something like that, but you may struggle with some parts later if you find you need to add more functionality. If you do choose to revisit FMTC and you need help again, please do get in touch.

**Reply by iBuaSpaz** [reply]
It doesnt really work anyways, adding 7.0.0 then it wont load the app, i remove the init code and from map, it loads fine.

**Reply by iBuaSpaz** [reply]
just fyi, i happens at 

FMTC.instance;

`I/flutter (16255): ╔══════════════════════════════════════════════════════════╗
I/flutter (16255): ║                   ISAR CONNECT STARTED                   ║
I/flutter (16255): ╟──────────────────────────────────────────────────────────╢
I/flutter (16255): ║           Open the link to connect to the Isar           ║
I/flutter (16255): ║          Inspector while this build is running.          ║
I/flutter (16255): ╟──────────────────────────────────────────────────────────╢
I/flutter (16255): ║ https://inspect.isar.dev/3.0.6-dev.0/#/52929/UlEUogFx2HI ║
I/flutter (16255): ╚══════════════════════════════════════════════════════════╝
E/flutter (16255): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: IsarError: Cannot open Environment: MdbxError (75): Value too large for defined data type`

**Reply by JaffaKetchup** [reply]
Yeah that's not supposed to happen, thanks for the report.

**Reply by JaffaKetchup** [reply]
@iBuaSpaz Was there a stack trace following that?

**Reply by iBuaSpaz** [reply]
no, this is all it writes.


√  Built build\app\outputs\flutter-apk\app-debug.apk.
Installing build\app\outputs\flutter-apk\app-debug.apk...          895ms
I/flutter (16578): ╔══════════════════════════════════════════════════════════╗
I/flutter (16578): ║                   ISAR CONNECT STARTED                   ║
I/flutter (16578): ╟──────────────────────────────────────────────────────────╢
I/flutter (16578): ║           Open the link to connect to the Isar           ║
I/flutter (16578): ║          Inspector while this build is running.          ║
I/flutter (16578): ╟──────────────────────────────────────────────────────────╢
I/flutter (16578): ║ https://inspect.isar.dev/3.0.6-dev.0/#/54281/XsCiCgW-cao ║
I/flutter (16578): ╚══════════════════════════════════════════════════════════╝
E/flutter (16578): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: IsarError: Cannot open Environment: MdbxError (75): Value too large for defined dq Quit (terminate the application on the device).

after that is just my normal debug output, so its all i see regarding your plugin

**Reply by JaffaKetchup** [reply]
Ok, thanks anyway, and I'll have to resolve that.

**Reply by iBuaSpaz** [reply]
could be a problem with ISAR if you really need it..

https://github.com/isar/isar/issues/1050

**Reply by JaffaKetchup** [reply]
Cheers. Are you running on an emulator?

**Reply by iBuaSpaz** [reply]
yes, pixel 5 api 30 emulator

**Reply by JaffaKetchup** [reply]
Ah, that explains it (see https://github.com/isar/isar/issues/860).

**Reply by iBuaSpaz** [reply]
this would be no issue for any mobiles, or it does not include the isar in release mode?

**Reply by JaffaKetchup** [reply]
All the reports only concern 32-bit Android emulators, so it looks like that's the only class of devices causing problems.

There's a possibility that older 32-bit real Android devices might run into the same issue, but I would've expected a bug report for that: so it doesn't look like an issue. I would double check, but I don't have a 32-bit device.

**Reply by iBuaSpaz** [reply]
I installed Pixed 6 x64 emulator api 31, and now it runs, until...

tileProvider: FMTC.instance('cache_maps').getTileProvider(),

then i just get alot of this:
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value
I/flutter ( 5617): Null check operator used on a null value

i comment out that line it shows the map (just curious if this works or not) :)

**Reply by JaffaKetchup** [reply]
Any stack trace for those null errors?

**Reply by JaffaKetchup** [reply]
And have you created that store?

**Reply by iBuaSpaz** [reply]
no, nothing.
im guessing all the null errors are tiles its checking, as nothing shows up on the map.

**Reply by JaffaKetchup** [reply]
Can you confirm that you've created the store? That's the most likely source of error here.

**Reply by iBuaSpaz** [reply]
> And have you created that store?

I was beginning to remove the code, and i removed that code i belive.

Now, this is more then easy for a newbie, where do i add the storemagement.create() again? im guessing your referring to that?

**Reply by iBuaSpaz** [reply]
I would really like to see this running just to see its possible to do this cache.

**Reply by iBuaSpaz** [reply]
I was beginning  to write a file just for how to get up and running with this fast, i guess im missing the Storemanagement.Create..

For quick up and running:

1) ----------
in your main.dart file:

import 'package:flutter_map_tile_caching/flutter_map_tile_caching.dart';

and modify your main() to include this

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await FlutterMapTileCaching.initialise();
  FMTC.instance;

  final store = FlutterMapTileCaching.instance('map_cache');
  await store.manage.createAsync();

  ........
}

2) ----------
in your dart file where you render the map, if not the main file, include this for caching.

import 'package:flutter_map_tile_caching/flutter_map_tile_caching.dart';

children: [
 TileLayer(
  tileProvider: FMTC.instance('map_cache').getTileProvider(),
  urlTemplate:
   ............
  ),
],

**Reply by JaffaKetchup** [reply]
Yeah, so you want to create your store probably in your `main()` method, below the intialisation. You may want to check if it exists first. The docs detail how to do this.

**Reply by JaffaKetchup** [reply]
Looks like you've just done it.

**Reply by JaffaKetchup** [reply]
I'm adding some improvements now to make these errors more obvious.

**Reply by iBuaSpaz** [reply]
with this it seems to work, isnt it supposed to show downloaded cache on isar? or is there a way to see if content shown is cached or downloaded?

**Reply by JaffaKetchup** [reply]
The Isar inspector is too basic for the setup here. You can check the caching is working by setting the tile provider to cache only mode, after panning over some tiles to cache them.

**Reply by iBuaSpaz** [reply]
okay, i did that test, it is saving them.

however, i build the apk in release mode and run on the phone, the phone ask if i want to clear the cache on first run, never opens, and then it just not running.

**Reply by JaffaKetchup** [reply]
Not sure what you mean by "asks to clear cache" then stops opening? Can you post a video or recording?

**Reply by iBuaSpaz** [reply]
Seems to be unrelated to the cache system, uncommenting it all still makes the error.

When launched the app first time, a popup came from bottom of screen before app opened asking if i wanted to clear cache, i said yes and then it stopped working.

**Reply by JaffaKetchup** [reply]
Never seen that before, but if it's not caused by fmtc, I can't do very much.

**Reply by iBuaSpaz** [reply]
just incase someone else searches this.
added this to android/app/build.gradle

    buildTypes {
        release {
            // TODO: Add your own signing config for the release build.
            // Signing with the debug keys for now, so `flutter run --release` works.
            signingConfig signingConfigs.debug
+            shrinkResources false
+            minifyEnabled false
        }
    }
    
and it runs, it seems :)

Just because im curious, IF i where to want to use this to save some time, and of course include that this uses your cache system and link to click, what would this cost me? im just a hobby programmer, and not even sure anyone would use my app so..

**Reply by JaffaKetchup** [reply]
Ah great, I might add that to the docs. In terms of licensing, please get in touch at fmtc@jaffaketchup.dev so we can discuss this privatley. If you could give me an idea of what your app is, and roughly how many users you forecast, I'm sure we can agree something. If it suits you better, I'm happy to do a per user cost basis, instead of yearly.

**Reply by JaffaKetchup** [reply]
@iBuaSpaz https://github.com/JaffaKetchup/flutter_map_tile_caching/pull/95 ;D
